Patch 14734 boolean sync validators

BobChao87 opened the pull request:
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**

Fix/clarification of how synchronous validators work.

**What is the current behavior? (You can also link to an open issue here)**
#14734 All values except `false` literal and `undefined` are treated as a pass. `undefined` is treated as pending, which makes no sense for synchronous operation.

**What is the new behavior (if this is a feature change)?**

All falsy value returns to a synchronous validator are treated as fails, including `undefined`.

**Does this PR introduce a breaking change?**

Yes, see commit message.

**Please check if the PR fulfills these requirements**
- [ x ] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [ x ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Other information**:

I believe a change to docs in unnecessary. We still expect `true` and `false` to be the primary returns and don't want to encourage otherwise. This is mostly clarifying a niche, undocumented, edge-case to behave more intuitively.

BobChao87 committed:
Demonstrate tests for issue #14734

BobChao87 committed:
fix(ngModel): treat synchronous validators as boolean always

Change synchronous validators to convert the return to boolean value.
Prevent unexpected behavior when returning `undefined`.

Closes #14734
BREAKING CHANGE: Previously, only a literal `false` return would resolve as the
synchronous validator failing. Now, all traditionally false JavaScript values
are treated as failing the validator, as one would naturally expect.

To migrate, make sure your synchronous validators are returning either a
literal `true` or a literal `false` value. For most code, we expect this to
already be the case. Only a very small subset of projects will affected.

Namely, anyone using `undefined` or any falsy value as a return will now see
their validation failing, whereas previously falsy values other than `undefined`
would have been seen as passing and `undefined` would have been seen as pending.

gkalpak commented on test/ng/directive/ngModelSpec.js:
```diff
@@ -847,6 +847,38 @@ describe('ngModel', function() {
       });
 
+      it('should treat all responses as boolean for synchronous validators', function() {
+        var curry = function(v) {
```
You don't need a new function for this. You can use the internal helper: `valueFn()`

BobChao87 replied:
Interesting, I'll make the switch. I was actually hoping to talk about the `curry()` anyways, because it's now used three times in this file (hence why I used it in the first place). I'll submit a separate PR to fix the others.

gkalpak replied:
Good catch! Feel free to change all other instances too.

gkalpak commented on test/ng/directive/ngModelSpec.js:
```diff
@@ -847,6 +847,38 @@ describe('ngModel', function() {
+          };
+        };
+
+        var tester = function(v, e) {
```
Nit: I would prefer a more descriptive name, such as `expectValid()`.

gkalpak reviewed (commented):
Apart from a couple of nitpicks, LGTM.

<sub>I would add a little more details to the commit message (regarding what was happening before) - can be done while merging.</sub>

BobChao87 committed:
fix(ngModel): treat synchronous validators as boolean always

Change synchronous validators to convert the return to boolean value.
Prevent unexpected behavior when returning `undefined`.

Closes #14734
BREAKING CHANGE: Previously, only a literal `false` return would resolve as the
synchronous validator failing. Now, all traditionally false JavaScript values
are treated as failing the validator, as one would naturally expect.

Specifically, the values `0` (the number zero), `null`, `NaN` and `''` (the
empty string) used to considered valid (passing) and they are now considered
invalid (failing). The value `undefined` was treated similarly to a pending
asynchronous validator, causing the validation to be pending. `undefined` is
also now considered invalid.

To migrate, make sure your synchronous validators are returning either a
literal `true` or a literal `false` value. For most code, we expect this to
already be the case. Only a very small subset of projects will be affected.

Namely, anyone using `undefined` or any falsy value as a return will now see
their validation failing, whereas previously falsy values other than `undefined`
would have been seen as passing and `undefined` would have been seen as pending.

BobChao87 committed:
fix(ngModel): treat synchronous validators as boolean always

Change synchronous validators to convert the return to boolean value.
Prevent unexpected behavior when returning `undefined`.

Closes #14734
BREAKING CHANGE: Previously, only a literal `false` return would resolve as the
synchronous validator failing. Now, all traditionally false JavaScript values
are treated as failing the validator, as one would naturally expect.

Specifically, the values `0` (the number zero), `null`, `NaN` and `''` (the
empty string) used to considered valid (passing) and they are now considered
invalid (failing). The value `undefined` was treated similarly to a pending
asynchronous validator, causing the validation to be pending. `undefined` is
also now considered invalid.

To migrate, make sure your synchronous validators are returning either a
literal `true` or a literal `false` value. For most code, we expect this to
already be the case. Only a very small subset of projects will be affected.

Namely, anyone using `undefined` or any falsy value as a return will now see
their validation failing, whereas previously falsy values other than `undefined`
would have been seen as passing and `undefined` would have been seen as pending.

BobChao87 committed:
fix(ngModel): treat synchronous validators as boolean always

Change synchronous validators to convert the return to boolean value.
Prevent unexpected behavior when returning `undefined`.

Closes #14734
BREAKING CHANGE: Previously, only a literal `false` return would resolve as the
synchronous validator failing. Now, all traditionally false JavaScript values
are treated as failing the validator, as one would naturally expect.

Specifically, the values `0` (the number zero), `null`, `NaN` and `''` (the
empty string) used to considered valid (passing) and they are now considered
invalid (failing). The value `undefined` was treated similarly to a pending
asynchronous validator, causing the validation to be pending. `undefined` is
also now considered invalid.

To migrate, make sure your synchronous validators are returning either a
literal `true` or a literal `false` value. For most code, we expect this to
already be the case. Only a very small subset of projects will be affected.

Namely, anyone using `undefined` or any falsy value as a return will now see
their validation failing, whereas previously falsy values other than `undefined`
would have been seen as passing and `undefined` would have been seen as pending.

BobChao87 commented:
Hopefully those changes to the commit message are in the direction you wanted.

gkalpak reviewed (approved):
LGTM

petebacondarwin commented:
Landed thanks.
